modules/4-graphql.livemd

```diff
@@ -2,89 +2,160 @@
 
-```elixir
+Resource exhaustion can occur inadvertently through legitimate use or could be triggered intentionally in a dos attack by a malicious actor who sends a large number or resource intensive requests to overload the application.  
 
-```
+The structure of GraphQL queries make it particularly succeptible to this type of attack as they can be crafted to perform long running and extensive operations, depending on the data being queried.
 
-<!-- livebook:{"branch_parent_index":4} -->
+In addition to strategies like implementing some form of rate limiting to protect APIs in general (see the section about rate limiting in the previous module), another approach to protect GraphQL from resource exhaustion involves anticipating the cost of a query and allocating resources based on known available capacity.  The next section introduces this approach.
 
 ## Cost Theory
 
 ### Description
 
-*TODO: Write Description*
+Resource intensive queries, like those where a Graphql query tries to traverse and then return a significant amount of highly nest data can cause a server/service to expend a significant amount of it's processing power and other resources.  These high cost queries can render a server and therefore the application useless.
 
-### Prevention
+One approach for implementing validation on incoming queries to determine their "cost" in terms of the resources the use.  Queries are defined by how much load they place on the server/service processing the request, allowing developers to plan for how best to manage resources.  This is a little like making a budget. 
 
-*TODO: Write Prevention*
+This approach also helps implement Rate limiting by establishing a query cost based on the type, operation, and expected performance of each unique GraphQL request for data, and by anticipating the load on the server.  
 
-### <span style="color:blue;">Example</span> / <span style="color:red;">Quiz</span>
+More details of this approach are described in the reference below.
 
-*TODO: Make Example or Quiz Question*
+### References
+https://shopify.engineering/rate-limiting-graphql-apis-calculating-query-complexity
 
-```elixir
+### GraphQl References
+https://cheatsheetseries.owasp.org/cheatsheets/GraphQL_Cheat_Sheet.html
+https://owasp.org/www-project-api-security/
+https://owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/12-API_Testing/01-Testing_GraphQL
+https://cheatsheetseries.owasp.org/cheatsheets/GraphQL_Cheat_Sheet.html
+https://owasp.org/www-project-api-security/  
+https://www.howtographql.com/advanced/4-security/
+(https://hexdocs.pm/hammer/tutorial.html)
```

Comment:
```suggestion
```

Reply:
Personally I think you've already done a great job linking out to relevant links in all the subsections throughout this module that this Resource chunk at the end is now unnecessary 🙂